File: src/crypto.ts
// Port of internal/excalidraw/crypto.go — uses Web Crypto API (crypto.subtle)

function base64urlDecode(s: string): Uint8Array {
  // base64url (no padding) → standard base64 → binary
  let b64 = s.replace(/-/g, '+').replace(/_/g, '/');
  while (b64.length % 4 !== 0) b64 += '=';
  const raw = atob(b64);
  const out = new Uint8Array(raw.length);
  for (let i = 0; i < raw.length; i++) out[i] = raw.charCodeAt(i);
  return out;
}

async function importKey(keyStr: string): Promise<CryptoKey> {
  const keyBytes = base64urlDecode(keyStr);
  return crypto.subtle.importKey('raw', keyBytes as BufferSource, { name: 'AES-GCM' }, false, ['decrypt']);
}

/**
 * Decrypts an Excalidraw static scene.
 * Wire format: 12-byte IV || ciphertext || 16-byte GCM tag
 */
export async function decrypt(keyStr: string, data: ArrayBuffer): Promise<Uint8Array> {
  if (data.byteLength < 12) throw new Error(`payload too short: ${data.byteLength} bytes`);
  const key = await importKey(keyStr);
  const iv = data.slice(0, 12);
  const ciphertext = data.slice(12);
  const plaintext = await crypto.subtle.decrypt({ name: 'AES-GCM', iv }, key, ciphertext);
  return new Uint8Array(plaintext);
}

/**
 * Decrypts a room broadcast where IV and ciphertext arrive as separate binary frames.
 */
export async function decryptGCM(keyStr: string, ciphertext: ArrayBuffer, iv: ArrayBuffer): Promise<Uint8Array> {
  const key = await importKey(keyStr);
  const plaintext = await crypto.subtle.decrypt({ name: 'AES-GCM', iv }, key, ciphertext);
  return new Uint8Array(plaintext);
}

File: cli/internal/excalidraw/crypto.go
package excalidraw

import (
	"crypto/aes"
	"crypto/cipher"
	"encoding/base64"
	"fmt"
)

// Decrypt decrypts AES-128-GCM data produced by Excalidraw's web client.
//
// Wire format: 12-byte IV || ciphertext || 16-byte GCM tag
// keyStr: base64url (no padding) encoded raw 128-bit key bytes
func Decrypt(keyStr string, data []byte) ([]byte, error) {
	if len(data) < 12 {
		return nil, fmt.Errorf("payload too short: %d bytes", len(data))
	}

	keyBytes, err := base64.RawURLEncoding.DecodeString(keyStr)
	if err != nil {
		return nil, fmt.Errorf("decode key: %w", err)
	}

	block, err := aes.NewCipher(keyBytes)
	if err != nil {
		return nil, fmt.Errorf("create cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("create GCM: %w", err)
	}

	iv := data[:gcm.NonceSize()]
	ciphertext := data[gcm.NonceSize():]

	plaintext, err := gcm.Open(nil, iv, ciphertext, nil)
	if err != nil {
		return nil, fmt.Errorf("decrypt: %w", err)
	}

	return plaintext, nil
}

// DecryptGCM decrypts AES-GCM data where the IV is provided separately.
// Used for room broadcasts where IV and ciphertext arrive as separate Socket.IO arguments.
func DecryptGCM(keyStr string, ciphertext, iv []byte) ([]byte, error) {
	keyBytes, err := base64.RawURLEncoding.DecodeString(keyStr)
	if err != nil {
		return nil, fmt.Errorf("decode key: %w", err)
	}

	block, err := aes.NewCipher(keyBytes)
	if err != nil {
		return nil, fmt.Errorf("create cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("create GCM: %w", err)
	}

	plaintext, err := gcm.Open(nil, iv, ciphertext, nil)
	if err != nil {
		return nil, fmt.Errorf("decrypt: %w", err)
	}

	return plaintext, nil
}
